### ISSUE TITLE
Rust detecting most types as "unknown" or outright getting types wrong all of the sudden

### ISSUE THREAD

**Issue by mattfj10** [rust-lang/rust-analyzer#12748]
**rust-analyzer version**: rust-analyzer version: 0.0.0 (5342f47f4 2022-07-09)

**rustc version**: rustc 1.61.0 (fe5b13d68 2022-05-18)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`): No relevant environment variables, but I do have Rust running inside a Nix environment. 

I have issues recently that appeared out of nowhere with Rust-Analyzer suddenly being unable to infer many types, often setting them to unknown or just outright getting them wrong, even in very simple cases. For example, I have a struct that's being parsed with clap, and originally the type inference worked when I did:

let data = example_struct::Parse()

originally it would correctly infer the type correctly as `:example_struct` but now it is `{unknown}`

When creating a new instance of a struct via an asynchronous function in one file it is able to infer the type:
`let result = result_struct::new().await` and will successfully label the type as `result_struct` but then in a similar file that uses that same identical line of code, it will simply say `{unknown}`

Or, when calling a random function I made called `restore` rust-analyzer was originally able to type it successfully as an anyhow result type of `Result<()>` but now rust-analyzer feels it is of `?Sized impl`
and even weirder the rust-analyzer will often triple list the types, so rather than it just listing the types once it will repeat it 3 times so it shows as `?Sized impl ?Sized impl ?Sized impl`

I thought perhaps this (even though it was initially working) out of nowhere loss in functionality was brought about by me using rust-analyzer with the local installation of Rust rather than the Nix one, but I tried launching VSCode from within the Nix repo, and even tried the nix environment switcher plugin and got the same strange analyzer results. I even tried running cargo clean and that also achieved nothing. Does this issue sound familiar to anyone?

**Comment by lnicola** [comment]
> No relevant environment variables, but I do have Rust running inside a Nix environment.

That might be very relevant.

Since you haven't provided a test case, can you reproduce it in a new project (so we can test)?

**Comment by lnicola** [comment]
If you're using the bundled binary (from the Code extension), we already do some `patchelf` magic to make it work on Nix. But I assume that doesn't help with the location of `rust-src`. Maybe you need to install that too somehow.

Anyway, I don't understand Nix. matklad is using it so it should work given the right setup.

**Comment by mattfj10** [comment]
> If you're using the bundled binary (from the Code extension), we already do some `patchelf` magic to make it work on Nix. But I assume that doesn't help with the location of `rust-src`. Maybe you need to install that too somehow.
> 
> Anyway, I don't understand Nix. matklad is using it so it should work given the right setup.

I just added an entry defining rust-src from the rustPkg and that seemed to sort it out. Not sure what initially caused it to freakout after it started working, but it seems its good to go now.

**Comment by lnicola** [comment]
Regarding the version number, where did you get the RA binary from?

**Comment by mattfj10** [comment]
> Regarding the version number, where did you get the RA binary from?

**Comment by mattfj10** [comment]
> Regarding the version number, where did you get the RA binary from?

I just got it from the Nix store. If I run the nix version locally in my terminal environment version number is given, just a date:
rust-analyzer 2022-05-17

Also, suddenly the problem of it not working when i run `code` from within the nix shell has returned but working when i run VSCode externally from nix, so it seems it wasn't fully solved RIP.

**Comment by lnicola** [comment]
> the Nix store

Oh great, does it have an app store now? :smile:

---

Did you set `rust-analyzer.server.path`? The extension comes with its own RA binary.

**Comment by mattfj10** [comment]
> > the Nix store
> 
> Oh great, does it have an app store now? 😄
> 
> Did you set `rust-analyzer.server.path`? The extension comes with its own RA binary.

Nah, store just means the local store of packages, this version of Rust though was pulled from the Nix repo, and I assume the `rust-analyzer` was as well.

**Comment by brianschnee** [comment]
## My issue
I just ran into a similar issue where everything being inferred by the LSP except for primitive types was read as `{unknown}`

## My fix
I was able to resolve this by checking my user `settings.json` on VSCode and removing this rule that was added: `"rust-analyzer.cargo.sysroot": null`.

### Navigation
1. open the command palette (`cmd + shift + p`)
2. search for `Preferences: Open User Settings (JSON)`
3. search for the rule above, comment it out and check for results

**Comment by VonTum** [comment]
I get the same issue as you describe it. 

For me it was having installed rustc and cargo through apt, instead of rustup. I figured out what was wrong because I ran `rustc --print sysroot` and got `/usr`

The fix was to uninstall everything, and install it through rustup.

**Comment by lnicola** [comment]
@VonTum you can also install `rust-src`, but I don't recommend using the distro toolchain. It's not supposed to be used.

**Comment by VonTum** [comment]
Ah yes, it seems to come with a base rustup install. 

The major thing with this is that it's very easy to miss, and it just makes the editor still do syntax highlighting, but type deduction and go-to-definition is really weak. I'd just been using the bad version for such a long time that I didn't really notice the lack of tooling. 

There really should be some kind of warning that it's not functioning fully, because otherwise people just accept that it doesn't work well.

**Comment by Veykril** [comment]
We used to report a message if we can't resolve `core` from the sysroot, but I might have messed that up with some rewrites in the past months.

**Comment by lnicola** [comment]
And it doesn't come with `rustup`, we just install it ourselves since https://github.com/rust-lang/rust-analyzer/pull/3206.

**Comment by Veykril** [comment]
No I just checked,we do emit the warning https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/project-model/src/sysroot.rs#L83 and https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/rust-analyzer/src/reload.rs#L160-L164
